Python/Default.sublime-keymap

```diff
@@ -1,25 +1,110 @@
 [
-    // Auto-pair quotes even after string modifiers.
-    // Copied over from the default bindings with modifications to `preceding_text`
-    // and an added selector condition.
+    // Suppress default auto-pairing bindings to avoid unwanted interference
+    { "keys": ["\""], "command": "insert", "args": {"characters": "\""}, "context":
+        [
+            { "key": "setting.auto_match_enabled" },
+            { "key": "selector", "operand": "source.python - string.quoted.double" },
+            { "key": "selection_empty", "match_all": true }
+        ]
+    },
+    // Auto-pair double quotes
+    // - not after escape \| -> \"|
+    // - not after quote "| -> ""|
+    // - after string modifiers: f| -> f"|"
+    // - in front of dots: |.join() -> "|".join()
+    // - in front of commas: |, -> "|",
+    // - in front of colons: |: -> "|":
+    // - in front of semicolons: |; -> "|";
     { "keys": ["\""], "command": "insert_snippet", "args": {"contents": "\"$0\""}, "context":
         [
-            { "key": "setting.auto_match_enabled", "operator": "equal", "operand": true },
-            { "key": "selection_empty", "operator": "equal", "operand": true, "match_all": true },
-            { "key": "following_text", "operator": "regex_contains", "operand": "^(?:\t| |\\)|]|\\}|>|$)", "match_all": true },
-            { "key": "preceding_text", "operator": "regex_contains", "operand": "(?i)\\b[bfru]+$", "match_all": true },
-            { "key": "selector", "operator": "equal", "operand": "source.python" },
-            { "key": "eol_selector", "operator": "not_equal", "operand": "string.quoted.double - punctuation.definition.string.end", "match_all": true }
+            { "key": "setting.auto_match_enabled" },
+            { "key": "selector", "operand": "source.python - string.quoted.double" },
+            { "key": "selection_empty", "match_all": true },
+            { "key": "preceding_text", "operator": "regex_contains", "operand": "(?i:^|[^\"\\w\\\\]|\\b[bfru]+)$", "match_all": true },
+            { "key": "following_text", "operator": "regex_contains", "operand": "^(?:\t| |\\)|]|\\}|\\.|,|:|;|$)", "match_all": true }
```

Comment:
Could we reasonably change this to match for any non-alphanumeric, i.e. `^(?!\w)`? While more general, that would simplify the regex a lot and I cannot reasonably think of a situation currently where this would be too eager. Same for single quotes, of course. Maybe even parens, brackets, and braces?

A similar thing is already done for `preceding_text`.

I would generally be happier if the non-Python-specific bindings (i.e. the non-string ones) would go into the global key bindings, but it doesn't seem like there is much interest for it. I shall try adding that to my config instead and see how it goes.

Reply:
As literly everything can be content of a quoted string, I am not convinced of such a common pattern. I am concerned about it producing too many false positives.